File: Day01/cpp/Day01/Day01.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <fstream>
#include <vector>

int main(int argc, char* argv[])
{
    std::string input1 = "D:\\Projects\\AdventOfCode\\2024\\AdventOfCode2024\\Day01\\Input1.txt";
    std::string input2 = "D:\\Projects\\AdventOfCode\\2024\\AdventOfCode2024\\Day01\\Input2.txt";
    // Build a read-Stream
    std::ifstream source;

    // Open the sample problem
    source.open(input1, std::ios_base::in);

    // Check if there was a problem opening the file
    if (!source)
    {
        std::cout << "Can't open file!";
        return 0;
    }

    // Vectors to hold all our numbers from each list
    std::vector<int> listOne;
    std::vector<int> listTwo;

    // Read in our puzzle input line by line
    for (std::string line; std::getline(source, line);)
    {
        // Make a stream for the line itself
        std::istringstream in(line);

        int firstNumber = 0;
        int secondNumber = 0;

        // Place the values from the list into our variables
        in >> firstNumber;
        in >> secondNumber;

        // Add out numbers to their respective list
        listOne.push_back(firstNumber);
        listTwo.push_back(secondNumber);
    }

    // Sort both lists into ascending order
    std::sort(listOne.begin(), listOne.end());
    std::sort(listTwo.begin(), listTwo.end());

    // Part 1
    {
        int answer = 0;

        // We can loop like this cause both lists are the same size (this would break if they weren't)
        for (int i = 0; i < listOne.size(); i++)
        {
            // Compare each pair and calculate their absolute difference (we don't care about negative distances)
            int difference = std::abs(listOne[i] - listTwo[i]);

            // And difference to our answer
            answer += difference;
        }
        
        printf("Answer: %d\n", answer);
    }

    // Part 2
    {
        int answer = 0;
        
        for (auto& i : listOne)
        {
            int val = i;
            int occurence = 0;
            
            for (auto & j : listTwo)
            {
                // If the number i from list one is the same as number j from list two then increment count
                if (i == j)
                {
                    occurence++;
                }
            }
            
            printf("%d occurs %d times.\n", val, occurence);
            answer += val * occurence;
        }

        printf("Answer: %d", answer);
    }
}

File: Day01/py/Day01.py
listOne, listTwo = [], []

input1 = "D:\\Projects\\AdventOfCode\\2024\\AdventOfCode2024\\Day01\\Input1.txt"
input2 = "D:\\Projects\\AdventOfCode\\2024\\AdventOfCode2024\\Day01\\Input2.txt"

## Open the puzzle input in read mode
with (open(input2, "r") as lines):

    ## For each line in the puzzle input
    for line in lines:

        ## Each line has two numbers so we can split to extract each number
        nums = line.split()

        ## Store each number as an int
        firstNumber = int(nums[0])
        secondNumber = int(nums[1])

        ## Add our numbers to their respective lists
        listOne.append(firstNumber)
        listTwo.append(secondNumber)

## Sort our lists in ascending order
listOne.sort()
listTwo.sort()
n = len(listOne)

## Part 1 #############################################
answer1 = 0

## We can loop like this cause both lists are the same size (this would break if they weren't)
for i in range(n):

    ## Compare each pair and calculate their absolute difference (we don't care about negative distances)
    difference = abs(listOne[i] - listTwo[i])
    answer1 += difference

print(answer1)

## Part 2 #############################################
answer2 = 0

## Loop over both lists and check how many times the numbers in listOne appear in listTwo
for i in listOne:
    occurrence = 0

    for j in listTwo:
        if i == j:
           occurrence += 1

    print(i, "occurs", occurrence, "times")
    answer2 += i * occurrence

print(answer2)
